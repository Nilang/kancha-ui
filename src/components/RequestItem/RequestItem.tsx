import React, { useState } from 'react'
import Container from '../Container/Container'
import Icon from '../Icon/Icon'
import Text, { TextTypes } from '../Text/Text'
import Radio from '../RadioBtn/RadioBtn'
import { TouchableHighlight } from 'react-native'
import { withTheme } from '../../theming'
import { RequestItemSelectable } from '../../types'
import S from 'string'

interface RequestItem {
  /**
   *  Pass a test ID to button
   */
  testID?: string
  /**
   * Remove bottom divider
   */
  last?: boolean

  /**
   * Remove bottom divider
   */
  dividerBottom?: boolean

  /**
   * Subtitle to show over main text
   */
  claimType: string

  /**
   *  Note to show on the right
   */
  itemNote?: string

  /**
   *  Credential property options
   */
  options: RequestItemSelectable[]

  /**
   *  The item being request is required
   */
  required?: boolean

  /**
   *  Return the selected item
   */
  onSelectItem: (id: string, claimType: string) => void

  /**
   *  Theme
   */
  closeAfterSelect?: boolean

  /**
   *  Theme
   */
  theme: any
}
/**
 * Simplified list otem for use in request flows
 */
const RequestItem: React.FC<RequestItem> = props => {
  const hasOptions = !props.required || props.options.length > 1
  const [options, updateSelected] = useState(props.options)
  const [optionsExpanded, toggleOptions] = useState(false)
  const [shared, toggleShared] = useState(true)
  const selected = options.find((i: RequestItemSelectable) => i.selected)
  const selectItem = (id: string) => {
    const updatedOptions = options.map(item => {
      return item.id === id ? { ...item, selected: true } : { ...item, selected: false }
    })
    updateSelected(updatedOptions)
    toggleShared(true)

    if (props.closeAfterSelect !== false) {
      toggleOptions(false)
    }

    props.onSelectItem(id, props.claimType)
  }
  const declineShare = () => {
    const updatedOptions = options.map(item => {
      return { ...item, selected: false }
    })
    updateSelected(updatedOptions)
    toggleShared(false)

    if (props.closeAfterSelect !== false) {
      toggleOptions(false)
    }

    props.onSelectItem('NOSHARE', props.claimType)
  }

  return (
    <TouchableHighlight
      onPress={() => toggleOptions(!optionsExpanded)}
      disabled={!hasOptions}
      testID={props.testID}
    >
      <Container background={'primary'} dividerBottom={!props.last || props.dividerBottom}>
        <Container
          paddingLeft={24}
          paddingTop={16}
          paddingBottom={16}
          flexDirection={'row'}
          alignItems={'center'}
        >
          <Container flex={1}>
            {props.claimType && (
              <Text type={TextTypes.SubTitle}>
                {S(props.claimType).capitalize().s}
                {props.required ? '*' : ''}
              </Text>
            )}
            <Container paddingTop={2}>
              <Text type={TextTypes.Body}>{selected ? selected.value : 'Not Shared'}</Text>
            </Container>
          </Container>
          {props.itemNote && (
            <Container paddingRight={32}>
              <Text type={TextTypes.ListItemNote}>{props.itemNote}</Text>
            </Container>
          )}
          {hasOptions && (
            <Container paddingRight={32}>
              <Icon icon={optionsExpanded ? props.theme.icons.UP : props.theme.icons.DOWN} />
            </Container>
          )}
        </Container>
        {hasOptions && optionsExpanded && (
          <Container marginLeft={8} paddingLeft={true} paddingBottom={true}>
            {!props.required && (
              <Container flexDirection={'row'}>
                <Container flex={1}>
                  <Radio disabled={props.required} selected={!shared} onPress={() => declineShare()}>
                    Do not share
                  </Radio>
                </Container>
              </Container>
            )}
            {options.map(item => {
              return (
                <Container key={item.id} flexDirection={'row'} alignItems={'center'}>
                  <Container flex={1}>
                    <Radio selected={item.selected} onPress={() => selectItem(item.id)}>
                      {item.value}
                    </Radio>
                  </Container>
                  <Container paddingRight={true}>
                    <Text type={TextTypes.ListItemNote}>{item.iss.shortId}</Text>
                  </Container>
                </Container>
              )
            })}
          </Container>
        )}
      </Container>
    </TouchableHighlight>
  )
}
export default withTheme(RequestItem)
