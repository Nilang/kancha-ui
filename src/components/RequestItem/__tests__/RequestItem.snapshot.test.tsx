import * as React from 'react'
import { render } from 'react-native-testing-library'
import RequestItem from '../RequestItem'

jest.mock('react-native-vector-icons/Ionicons', () => 'Icon')

const singleOption = [
  {
    id: 'TEST_ID',
    iss: { shortId: 'Serto Verified', did: '0xfksksdk' },
    type: 'Test property',
    value: 'Test value',
    selected: true,
  },
]

const multiOptions = [
  {
    id: 'TEST_ID_01',
    iss: { shortId: 'Serto Verified', did: '0xfksksdk' },
    type: 'Test property',
    value: 'Test value 01',
    selected: true,
  },
  {
    id: 'TEST_ID_02',
    iss: { shortId: 'Serto Verified', did: '0xfksksdk' },
    type: 'Test property',
    value: 'Test value 02',
    selected: false,
  },
]

describe('Component(snapshot): RequestItem', () => {
  it('should render required with single option', () => {
    const onSelect = jest.fn()
    const tree = render(
      <RequestItem claimType={'Test'} options={singleOption} onSelectItem={onSelect} />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render un-required with single option', () => {
    const onSelect = jest.fn()
    const tree = render(
      <RequestItem required={false} claimType={'Test'} options={singleOption} onSelectItem={onSelect} />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render required with multi options', () => {
    const onSelect = jest.fn()
    const tree = render(
      <RequestItem
        required={true}
        claimType={'Subtitle'}
        itemNote={'Test Note'}
        last={true}
        options={multiOptions}
        onSelectItem={onSelect}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render un-required with multi options', () => {
    const onSelect = jest.fn()
    const tree = render(
      <RequestItem
        required={false}
        claimType={'Subtitle'}
        itemNote={'Test Note'}
        last={true}
        options={multiOptions}
        onSelectItem={onSelect}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
