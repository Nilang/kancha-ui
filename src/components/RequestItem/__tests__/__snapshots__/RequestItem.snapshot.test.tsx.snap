// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshot): RequestItem should render with all default props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": "row",
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": 16,
      "paddingHorizontal": undefined,
      "paddingLeft": 24,
      "paddingRight": undefined,
      "paddingTop": 16,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#BBBBBB",
          "fontSize": 14,
        }
      }
    >
      Subtitle
    </Text>
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": 2,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "fontSize": 16,
            "lineHeight": 22,
          }
        }
      >
        Hello Test
      </Text>
    </View>
  </View>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": 32,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#BBBBBB",
          "fontSize": 15,
        }
      }
    >
      Test Note
    </Text>
  </View>
</View>
`;

exports[`Component(snapshot): RequestItem should render with default props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderBottomColor": "#CCCCCC",
      "borderBottomWidth": 0.5,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": "row",
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": 16,
      "paddingHorizontal": undefined,
      "paddingLeft": 24,
      "paddingRight": undefined,
      "paddingTop": 16,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": 2,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "fontSize": 16,
            "lineHeight": 22,
          }
        }
      >
        Hello Test
      </Text>
    </View>
  </View>
</View>
`;
